opentelemetry-api/src/opentelemetry/metrics/instrument.py

```diff
@@ -0,0 +1,204 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=too-many-ancestors
+
+# type: ignore
+
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+from re import compile as compile_
+from threading import RLock
+from types import GeneratorType
+
+from opentelemetry.metrics.measurement import Measurement
+
+_logger = getLogger(__name__)
+
+
+class Instrument(ABC):
+
+    _name_regex = compile_(r"[a-zA-Z][-.\w]{0,62}")
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def unit(self):
+        return self._unit
+
+    @property
+    def description(self):
+        return self._description
+
+    @abstractmethod
+    def __init__(self, name, *args, unit="", description="", **kwargs):
+        self._lock = RLock()
+
+        if name is None or self._name_regex.fullmatch(name) is None:
+            _logger.error("Invalid instrument name %s", name)
+
+        else:
+            self._name = name
+
+        if unit is None:
+            self._unit = ""
+        elif len(unit) > 63:
+            _logger.error("unit must be 63 characters or shorter")
+
+        elif any(ord(character) > 127 for character in unit):
+            _logger.error("unit must only contain ASCII characters")
+        else:
+            self._unit = unit
+
+        if description is None:
+            description = ""
+
+        self._description = description
+
+
+class Synchronous(Instrument):
+    pass
+
+
+class Asynchronous(Instrument):
+    @abstractmethod
+    def __init__(
+        self, name, callback, *args, unit="", description="", **kwargs
+    ):
+        super().__init__(name, *args, unit=unit, description="", **kwargs)
+        with self._lock:
+
+            if not isinstance(callback, GeneratorType):
+                _logger.error("callback must be a generator")
```

Comment:
If the user passes a generator here, how do you reset that generator for the next `collect()` round? It seems like what you want is a generator *factory*.

My only concern with generator use here is that makes the most common use case a little awkward, where folks aren't expecting to need a generator for a callback that returns only a single measurement. Could we support both generator and not generator callbacks?

Reply:
> If the user passes a generator here, how do you reset that generator for the next `collect()` round? It seems like what you want is a generator _factory_.

You can't/don't reset a generator, that's the point of using one. From [this](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#asynchronous-gauge-creation) part of the spec:

_The API SHOULD provide some way to pass state to the callback. Individual language client can decide what is the idiomatic approach (e.g. it could be an additional parameter to the callback function, or captured by the lambda closure, or something else)._

The Python idiomatic approach for a Python callable that has a state _is_ the generator.

> My only concern with generator use here is that makes the most common use case a little awkward, where folks aren't expecting to need a generator for a callback that returns only a single measurement.

Sorry, but what do you mean with awkward? Python generators serve exactly the purpose that we want for callbacks that can have a state. It would be awkward not to use them.

> Could we support both generator and not generator callbacks?

We could check the type of the callback and call use `next` or parentheses. :shrug:

Reply:
@ocelotl can you please show a code example of how to create an async instrument with a generator. I don't think I understand from just reading the code.

Reply:
> @ocelotl can you please show a code example of how to create an async instrument with a generator. I don't think I understand from just reading the code.

Sure, just above, right [here](https://github.com/open-telemetry/opentelemetry-python/pull/1887#discussion_r711067992) :v: